src/lib/storage/dictionary_column.hpp

```diff
@@ -0,0 +1,192 @@
+// types (uint8_t, uint16_t) since after a down-cast INVALID_VALUE_ID will look like their numeric_limit::max()
+constexpr ValueID INVALID_VALUE_ID{std::numeric_limits<ValueID::base_type>::max()};
+
+// Dictionary is a specific column type that stores all its values in a vector
+template <typename T>
+class DictionaryColumn : public BaseColumn {
+ public:
+  /**
+   * Creates a Dictionary column from a given value column.
+   */
+  explicit DictionaryColumn(const std::shared_ptr<BaseColumn>& base_column);
+
+  // SEMINAR INFORMATION: Since most of these methods depend on the template parameter, you will have to implement
+  // the DictionaryColumn in this file. Replace the method signatures with actual implementations.
+
+  // return the value at a certain position. If you want to write efficient operators, back off!
+  const AllTypeVariant operator[](const size_t i) const override;
+
+  // return the value at a certain position.
+  const T get(const size_t i) const;
+
+  // dictionary columns are immutable
+  void append(const AllTypeVariant&) override;
+
+  // returns an underlying dictionary
+  std::shared_ptr<const std::vector<T>> dictionary() const;
+
+  // returns an underlying data structure
+  std::shared_ptr<const BaseAttributeVector> attribute_vector() const;
+
+  // return the value represented by a given ValueID
+  const T& value_by_value_id(ValueID value_id) const;
+
+  // returns the first value ID that refers to a value >= the search value
+  // returns INVALID_VALUE_ID if all values are smaller than the search value
+  ValueID lower_bound(const T value) const;
+
+  // same as lower_bound(T), but accepts an AllTypeVariant
+  ValueID lower_bound(const AllTypeVariant& value) const;
+
+  // returns the first value ID that refers to a value > the search value
+  // returns INVALID_VALUE_ID if all values are smaller than or equal to the search value
+  ValueID upper_bound(const T value) const;
+
+  // same as upper_bound(T), but accepts an AllTypeVariant
+  ValueID upper_bound(const AllTypeVariant& value) const;
+
+  // return the number of unique_values (dictionary entries)
+  size_t unique_values_count() const;
+
+  // return the number of entries
+  size_t size() const override;
+
+ protected:
+  std::shared_ptr<BaseAttributeVector> _get_fitted_vector(const size_t size);
+
+  std::shared_ptr<std::vector<T>> _dictionary;
+  std::shared_ptr<BaseAttributeVector> _attribute_vector;
+};
+
+template <typename T>
+DictionaryColumn<T>::DictionaryColumn(const std::shared_ptr<BaseColumn>& base_column)
+    : _dictionary(nullptr), _attribute_vector(nullptr) {
+  DebugAssert(base_column, "No base column provided");
+
+  const auto value_column = std::dynamic_pointer_cast<ValueColumn<T>>(base_column);
+
+  DebugAssert(value_column, "Bad value column provided");
+
+  auto sorted_values = value_column->values();
+  std::sort(sorted_values.begin(), sorted_values.end());
```

Comment:
Actually I'm having second thoughts about this one. Since this is not `auto& sorted_values = ...`, you might be creating a copy on the stack. This is also not desired, supposedly it should go straight into the target vector.